File: CRC-32.py
import binascii
from bitstring import BitArray

class CRC32:
    def __init__(self):
        self.polynomial = 0x04C11DB7

    def calculate(self, bit_string):
        # Rellenar con ceros a la izquierda hasta que la longitud sea un múltiplo de 8
        while len(bit_string) % 8 != 0:
            bit_string = '0' + bit_string

        bit_array = BitArray(bin=bit_string)
        crc = binascii.crc32(bit_array.bytes) & 0xFFFFFFFF
        return format(crc, '032b')

class Sender:
    def __init__(self, crc):
        self.crc = crc

    def send(self, bit_string):
        crc = self.crc.calculate(bit_string)
        print("Mensaje original: " + bit_string)
        print("CRC calculado: " + crc)
        return bit_string + crc

class Receiver:
    def __init__(self, crc):
        self.crc = crc

    def receive(self, bit_string):
        data, crc_received = bit_string[:-32], bit_string[-32:]
        crc_calculated = self.crc.calculate(data)

        print("Mensaje original: " + data)
        print("CRC recibido: " + crc_received)
        print("CRC calculado: " + crc_calculated)

        if crc_calculated == crc_received:
            return "No se detectaron errores."
        else:
            # Comprobamos cada bit individualmente
            for i in range(len(data)):
                bit = data[i]
                modified_data = data[:i] + ('1' if bit == '0' else '0') + data[i+1:]
                if self.crc.calculate(modified_data) == crc_received:
                    return f"Se detectó un error en el bit {i+1}: la trama se descarta"
            return "Se detectaron errores desconocidos: la trama se descarta"

# Ejemplo de uso
crc = CRC32()
sender = Sender(crc)
receiver = Receiver(crc)

# Trama sin manipular
data1 = "110101"
sent_data1 = sender.send(data1)
print("\nEmisor envía: " + sent_data1)
print(receiver.receive(sent_data1))

data2 = "101010"
sent_data2 = sender.send(data2)
print("\nEmisor envía: " + sent_data2)
print(receiver.receive(sent_data2))

data3 = "11110000"
sent_data3 = sender.send(data3)
print("\nEmisor envía: " + sent_data3)
print(receiver.receive(sent_data3))

# Tramas con un bit modificado
sent_data1_modified = "010101" + sent_data1[6:]
print("\nEmisor envía: " + sent_data1_modified)
print(receiver.receive(sent_data1_modified))

sent_data2_modified = "001010" + sent_data2[6:]
print("\nEmisor envía: " + sent_data2_modified)
print(receiver.receive(sent_data2_modified))

sent_data3_modified = "01110000" + sent_data3[8:]
print("\nEmisor envía: " + sent_data3_modified)
print(receiver.receive(sent_data3_modified))

# Tramas con dos bits modificados
print("\nTramas con dos bits modificados")
sent_data1_modified = "010101" + sent_data1[6:]
print("\nEmisor envía: " + sent_data1_modified)
print(receiver.receive(sent_data1_modified))

sent_data2_modified = "001010" + sent_data2[6:]
print("\nEmisor envía: " + sent_data2_modified)
print(receiver.receive(sent_data2_modified))

sent_data3_modified = "01110000" + sent_data3[8:]
print("\nEmisor envía: " + sent_data3_modified)
print(receiver.receive(sent_data3_modified))

print("\nTrama modificada para que el algoritmo no detecte el error")

sent_data1_modified = "010101" + sent_data1[6:]
sent_data1_modified_undetectable = "010101" + sender.send("010101")[6:]
print("\nEmisor envía: " + sent_data1_modified_undetectable)
print(receiver.receive(sent_data1_modified_undetectable))

File: CRC-32.js
const crc32 = require('buffer-crc32');

class CRC32 {
    constructor() {
        this.polynomial = 0x04C11DB7;
    }

    calculate(bitString) {
        while (bitString.length % 8 !== 0) {
            bitString = '0' + bitString;
        }

        let bitArray = Buffer.from(bitString, 'binary');
        let crc = crc32.unsigned(bitArray);
        return crc.toString(2).padStart(32, '0');
    }
}

class Sender {
    constructor(crc) {
        this.crc = crc;
    }

    send(bitString) {
        let crc = this.crc.calculate(bitString);
        console.log("Mensaje original: " + bitString);
        console.log("CRC calculado: " + crc);
        return bitString + crc;
    }
}

class Receiver {
    constructor(crc) {
        this.crc = crc;
    }

    receive(bitString) {
        let data = bitString.slice(0, -32);
        let crcReceived = bitString.slice(-32);
        let crcCalculated = this.crc.calculate(data);

        console.log("Mensaje original: " + data);
        console.log("CRC recibido: " + crcReceived);
        console.log("CRC calculado: " + crcCalculated);

        if (crcCalculated === crcReceived) {
            return "No se detectaron errores.";
        } else {
            for (let i = 0; i < data.length; i++) {
                let bit = data[i];
                let modifiedData = data.slice(0, i) + (bit === '0' ? '1' : '0') + data.slice(i + 1);
                if (this.crc.calculate(modifiedData) === crcReceived) {
                    return `Se detectó un error en el bit ${i + 1}: la trama se descarta`;
                }
            }
            return "Se detectaron errores desconocidos: la trama se descarta";
        }
    }
}

// Ejemplo de uso
let crc = new CRC32();
let sender = new Sender(crc);
let receiver = new Receiver(crc);

// Trama sin manipular
let data1 = "110101";
let sentData1 = sender.send(data1);
console.log("\nEmisor envía: " + sentData1);
console.log(receiver.receive(sentData1));

let data2 = "101010";
let sentData2 = sender.send(data2);
console.log("\nEmisor envía: " + sentData2);
console.log(receiver.receive(sentData2));

let data3 = "11110000";
let sentData3 = sender.send(data3);
console.log("\nEmisor envía: " + sentData3);
console.log(receiver.receive(sentData3));

// Tramas con un bit modificado
let sentData1Modified = "010101" + sentData1.slice(6);
console.log("\nEmisor envía: " + sentData1Modified);
console.log(receiver.receive(sentData1Modified));

let sentData2Modified = "001010" + sentData2.slice(6);
console.log("\nEmisor envía: " + sentData2Modified);
console.log(receiver.receive(sentData2Modified));

let sentData3Modified = "01110000" + sentData3.slice(8);
console.log("\nEmisor envía: " + sentData3Modified);
console.log(receiver.receive(sentData3Modified));

// Tramas con dos bits modificados
console.log("\nTramas con dos bits modificados");
sentData1Modified = "010101" + sentData1.slice(6);
console.log("\nEmisor envía: " + sentData1Modified);
console.log(receiver.receive(sentData1Modified));

sentData2Modified = "001010" + sentData2.slice(6);
console.log("\nEmisor envía: " + sentData2Modified);
console.log(receiver.receive(sentData2Modified));

sentData3Modified = "01110000" + sentData3.slice(8);
console.log("\nEmisor envía: " + sentData3Modified);
console.log(receiver.receive(sentData3Modified));

console.log("\nTrama modificada para que el algoritmo no detecte el error");

sentData1Modified = "010101" + sentData1.slice(6);
let sentData1ModifiedUndetectable = "010101" + sender.send("010101").slice(6);
console.log("\nEmisor envía: " + sentData1ModifiedUndetectable);
console.log(receiver.receive(sentData1ModifiedUndetectable));
